TER-339 make harvest modules faster

kanaksinghal opened the pull request:
This change attempts to make the harvest process faster by allowing user to group compatible modules and run `terraform init` by group instead of by-module
This change is also backward compatible. Therefore, when groups are not set, the behaviour is same as before.

petmal commented:
I wonder if it would be easier (i.e. less complexity) if we just allow an option to run single init for the entire file. By default there would be init per module, but with the option flag everything would be checked out into single dir.

The farm yaml file is the grouping mechanism. Instead of defining group property on module you would just gather "compatible" modules in a single yaml file.

I feel like that is going to be easier to maintain than groups inside a single large yaml file.

kanaksinghal commented:
> I wonder if it would be easier (i.e. less complexity) if we just allow an option to run single init for the entire file. By default there would be init per module, but with the option flag everything would be checked out into single dir.
> 
> 
> 
> The farm yaml file is the grouping mechanism. Instead of defining group property on module you would just gather "compatible" modules in a single yaml file.
> 
> 
> 
> I feel like that is going to be easier to maintain than groups inside a single large yaml file.

I was aiming for backward compatibility and minimum change. However, I agree that the yaml format could have better reflected the eventual goal, which is to have one FarmModuleGroup struct per file. If we split the files so that each file represents a group, the current file would fail validation as it cannot contain different values of the export flag in one group. Additionally, the modules in the current file do not work together during initialization.

For now, I prioritized backward compatibility because any other approach would have immediate effects. In the future, when we transition to using groups, we can consider the most efficient format. Moreover, the current approach allows for the exclusion of a group, which might be useful for scanning old templates for mapping. If we split into files, it would mean that a user can’t do this without creating a file for each module. We will only know if this is necessary when we encounter it.

One other issue I’ve been grappling with is how to explain to a user that this group is solely intended for tf init efficiency and not for categorizing the modules. When I was working on the farm repository, I initially created three groups there - public, xa, templates. Later, I realized that public and xa works together, so I switched to g1, g2 for more efficiency. Thus, this shouldn’t be used for categorization purposes.

rajatsharma111 commented:
@kanaksinghal I have raised a Automation test PR - https://github.com/cldcvr/terrarium-api-automation/pull/20
